kratos/includes/chunk.h

```diff
@@ -94,68 +90,64 @@ namespace Kratos
       ///@name Operations
       ///@{
 	  void Initialize() {
-		  mOwnerThread = OpenMPUtils::ThisThread();  // initialization can change the owner thread
-		  std::size_t block_size_after_alignment = GetBlockSize(mBlockSizeInBytes);
-		  mpData = new DataType[DataSize()];
-  		  SetFirstAvailableBlockIndex(0);
-		  for (auto i_thread = 0; i_thread < OpenMPUtils::GetCurrentNumberOfThreads(); i_thread++)
-			  SetNumberOfAvailableBlocks(0, i_thread);
-		  SetNumberOfAvailableBlocks(AllocatableDataSize() / block_size_after_alignment);
-
-		  SizeType i_block = 0;
-		  for (auto p = GetData(); i_block < GetNumberOfBlocks(); p += block_size_after_alignment)
-			  *p = ++i_block;
-
-		  KRATOS_DEBUG_CHECK_EQUAL(i_block, GetNumberOfAvailableBlocks());
+		  const std::size_t data_size = DataSize();
+		  mpData = new BlockType[data_size];
+		  mpEnd = mpData + data_size;
+  		  mFirstAvailableBlock = mpData;
+		  mNumberOfAvailableBlocks = AllocatableDataSize() / mBlockSizeAfterAlignment;
+		  mpUninitializedMemory = mpData;
+		  
+
+		  *mpData = 0; // The first entry of the link list to the next one
 	  }
 
 	  /// This function does not throw and returns zero if cannot allocate
 	  void* Allocate() {
-		  KRATOS_DEBUG_CHECK_EQUAL(mOwnerThread, OpenMPUtils::ThisThread()); // Allocate should be called only by owner thread
 		  KRATOS_DEBUG_CHECK_NOT_EQUAL(mpData, nullptr);
 
-		  if (GetTotalNumberOfAvailableBlocks() == 0)
+		  if (IsFull())
 			  return nullptr;
 
-		  if (GetNumberOfAvailableBlocks() == 0) // Time to get blocks from other threads
-		  {
-			  lock();
-			  for (auto i_thread = 0; i_thread < OpenMPUtils::GetCurrentNumberOfThreads(); i_thread++)
-				  if (GetNumberOfAvailableBlocks(i_thread) != 0) {
-					  SetFirstAvailableBlockIndex(GetFirstAvailableBlockIndex(i_thread));
-					  SetNumberOfAvailableBlocks(GetNumberOfAvailableBlocks(i_thread));
-					  SetNumberOfAvailableBlocks(0, i_thread);
-					  break;
-				  }
-			  unlock();
-		  }
-		  DataType * p_result = GetData() + (GetFirstAvailableBlockIndex() * GetBlockSize(mBlockSizeInBytes));
-		  KRATOS_DEBUG_CHECK(Has(p_result));
-		  SetFirstAvailableBlockIndex(*p_result);
-		  SetNumberOfAvailableBlocks(GetNumberOfAvailableBlocks()-1);
-
+			lock();  
+			BlockType * p_result = mFirstAvailableBlock;
+			mFirstAvailableBlock = (p_result >= mpUninitializedMemory) ? mFirstAvailableBlock + mBlockSizeAfterAlignment : (BlockType*)*p_result;
+			// if(p_result >= mpUninitializedMemory) {
+			// 	mFirstAvailableBlock += mBlockSizeAfterAlignment;
+			// 	if(mpUninitializedMemory < mpEnd) {
+			// 		mpUninitializedMemory += mBlockSizeAfterAlignment;
```

Comment:
Remove commented code?

Reply:
I leaved it as it explains the logic of the above line. (with a small change which I leave to test)